Compute tria intersections

fdrmrc opened the pull request:
- Fixed numbering of vertices in rectangle-rectangle intersection 0fbcacd13f20c660dc1f885d66783f0f289eea6d

 - Added new function `intersect_and_get_quads()`. It returns a ```std::vector<
  std::tuple<typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,Quadrature<spacedim>>>```

- Extended namespace `dealii::NonMatching` so that it contains `compute_intersection` (that works for cells only) and the new functions for triangulations

luca-heltai commented on include/compute_intersections_of_tria.h:
```diff
@@ -0,0 +1,54 @@
+    const unsigned int                      degree);
+
+}
+
+
+
```
```suggestion
```

luca-heltai commented on include/compute_intersections_of_tria.h:
```diff
@@ -0,0 +1,54 @@
+#include <vector>
+
+#include "compute_intersection_of_cells.h"
+
+
```
```suggestion
```

luca-heltai commented on include/compute_intersections_of_tria.h:
```diff
@@ -0,0 +1,54 @@
+    typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,
+    typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,
+    dealii::Quadrature<spacedim>>>
+  intersect_and_get_quads(
```
I would rather call also this function `compute_intersection`, leveraging C++ overloading capabilities.

Can you add a documentation for the function, and keep both functions in the same header file?

fdrmrc commented:
Thanks Luca for the review. I added operator overloading and documentation for `compute_intersection`. I renamed the files and fixed the #includes in the `gtests` folder.

luca-heltai commented:
Documentation is still missing. Would you mind adding some documentation? Remember: someone who has never seen this before, should understand what it does, and how to use it. :)

luca-heltai commented on source/compute_intersections.cc:
```diff
@@ -0,0 +1,387 @@
+
+namespace dealii::NonMatching
+{
+  /**
+   * @brief Intersect `cell0` and `cell1` and construct a `Quadrature<spacedim>` of degree `degree``
+   *        over the intersection, i.e. in the real space. Mappings for both
+   * cells are in `mapping0` and `mapping1`, respectively.
+   *
+   * @tparam dim0
+   * @tparam dim1
+   * @tparam spacedim
+   * @param cell0 A `cell_iteratator` to the first cell
+   * @param cell1 A `cell_iteratator` to the first cell
+   * @param degree The degree of the `Quadrature` you want to build there
+   * @param mapping0 The `Mapping` object describing the first cell
+   * @param mapping1 The `Mapping` object describing the second cell
+   * @return Quadrature<spacedim>
+   */
```
This should go in the header file, not in the source file. :)

luca-heltai commented on source/compute_intersections.cc:
```diff
@@ -0,0 +1,387 @@
+  }
+
+
+  /**
+   * @brief Given two triangulations cached inside `GridTools::Cache` objects, compute all intersections between the two
+   * and return a vector where each entry is a tuple containing iterators to the
+   * respective cells and a `Quadrature<spacedim>` formula to integrate over the
+   * intersection.
+   *
+   * @tparam dim0 Intrinsic dimension of the immersed grid
+   * @tparam dim1 Intrinsic dimension of the ambient grid
+   * @tparam spacedim
+   * @param space_cache
+   * @param immersed_cache
+   * @param degree Degree of the desired quadrature formula
+   * @return std::vector<std::tuple<
+   * typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,
+   * typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,
+   * dealii::Quadrature<spacedim>>>
+   */
```
Same here. The documentation goes together with the declaration, not in the cc file, but in the h file.

luca-heltai commented:
If you need help with the rebase, let me know, and I'll do it for you (or explain to you how to do it, so you know it next time).

luca-heltai commented on source/compute_intersections.cc:
```diff
@@ -0,0 +1,375 @@
+  }
+
+} // namespace dealii::NonMatching
+
```
Add an `#else`  and `#endif` here, where you have the default implementations that simply throws exceptions saying that you need CGAL, and then instantiate all of the functions irrespective of wether CGAL exists or not. This makes sure you don't get errors in user codes at link time, but just at run time.

fdrmrc commented:
- I added here `create_coupling_sparsity_pattern_with_exact_intersections()` . As you suggested, this depends on the output of `compute_intersection()`, which is a vector of tuples.
- `compute_intersection` has been improved by avoiding to store the intersected cells in a `std::set`, so that only a double for over the trees is needed. 

This sparsity pattern has been checked in `reduced_lagrange.cc`, and runs fine

luca-heltai commented on source/compute_intersections.cc:
```diff
@@ -0,0 +1,372 @@
+
+
+#else
+
+Assert(DEAL_II_WITH_CGAL,
+       ExcMessage("This function needs CGAL to be installed, "
+                  "but cmake could not find it."));
```
```suggestion
template <int dim0, int dim1, int spacedim>
  std::vector<std::tuple<
    typename dealii::Triangulation<dim0, spacedim>::active_cell_iterator,
    typename dealii::Triangulation<dim1, spacedim>::active_cell_iterator,
    dealii::Quadrature<spacedim>>>
  compute_intersection(const GridTools::Cache<dim0, spacedim> &,
                       const GridTools::Cache<dim1, spacedim> &,
                       const unsigned int                     )
  {
Assert(false,
       ExcMessage("This function needs CGAL to be installed, "
                  "but cmake could not find it."));
                  return {};
}
     
                 template <int dim0, int dim1, int spacedim>
  dealii::Quadrature<spacedim>
  compute_intersection(
    const typename dealii::Triangulation<dim0, spacedim>::cell_iterator &,
    const typename dealii::Triangulation<dim1, spacedim>::cell_iterator &,
    const unsigned int                                                  ,
    const dealii::Mapping<dim0, spacedim> &,
    const dealii::Mapping<dim1, spacedim> &)
  {
               Assert(false,
       ExcMessage("This function needs CGAL to be installed, "
                  "but cmake could not find it."));
                  return {};
                  }
```
